### ISSUE TITLE
Is it possible to control the speed of the speech?

### ISSUE THREAD

**Issue by godspirit00** [yl4579/StyleTTS#3]
Hello,
It truly is an impressive work!
I wonder if it is possible to control the speed of the output speech.
Thank you!

**Comment by yl4579** [comment]
Yes, just scale the duration predicted. Simply change the line 
```python
duration = model.predictor.duration_proj(x)
```
in [Inference_LJSpeech.ipynb](https://github.com/yl4579/StyleTTS/blob/main/Demo/Inference_LJSpeech.ipynb) with
```python
duration = model.predictor.duration_proj(x) / speed
```
where `speed ` is your scaling factor, for example, 1.25 is 125% faster.

You can also change the speed more naturally by giving slower or faster reference audio, which is a main selling point of StyleTTS.

**Comment by godspirit00** [comment]
Thanks for the reply!
Just another question, is a reference speech always needed to do synthesis?

**Comment by yl4579** [comment]
Yes, it is always necessary. But you can precompute a style from a reference and add some noise to the style to synthesize diverse speech.

**Comment by sinhprous1** [comment]
Hi @yl4579 , could we also control the pitch of the generated audio, i.e: generate different audio variations of a speaker from single text?

**Comment by vik-rant** [comment]
@yl4579 , any way to control the pitch? changing the predicted F0 values before sending to the decoder seems to make no difference.

**Comment by yl4579** [comment]
@vik-rant @sinhprous1 Change the F0 curve should change the pitch, can you give me some examples that changing F0 curves don't change the pitch?

**Comment by vik-rant** [comment]
@yl4579 , there is no difference in output when the predicted F0 is modified, even with large values like +/-100000
```
        F0_pred, N_pred = model.predictor.F0Ntrain(en, s)
        F0_pred += 100
        print('F0:')
        print(F0_pred.shape)
        print(F0_pred)

        out = model.decoder((t_en @ pred_aln_trg.unsqueeze(0).to(device)), 
                                F0_pred, N_pred, ref.squeeze().unsqueeze(0))
```
sharing some generated samples with different F0 values, but they are all basically the same:
https://drive.google.com/drive/folders/1au9JGH-WhGTxkeErTT0wNNBWU3UfBEAD?usp=sharing

**Comment by vik-rant** [comment]
sharing some samples with pitch shifted by various values:
https://drive.google.com/drive/folders/1au9JGH-WhGTxkeErTT0wNNBWU3UfBEAD?usp=sharing
```
        F0_pred, N_pred = model.predictor.F0Ntrain(en, s)
        F0_pred += 100
        print('F0:')
        print(F0_pred.shape)
        print(F0_pred)

        out = model.decoder((t_en @ pred_aln_trg.unsqueeze(0).to(device)), 
                                F0_pred, N_pred, ref.squeeze().unsqueeze(0))
```

**Comment by vik-rant** [comment]
@yl4579 , got it, thanks.

**Comment by chiaki-luo** [comment]
@yl4579 Dear author, I have some questions about the length of output speech. During training, the time dim can be fixed, but the length may be different when inference. When I directly listen the mel produced in training stage, it sounds worse than inference. I wonder why it works, and whether StyleTTS-VC model can perform better by using output of inference as ground truth.

**Comment by yl4579** [comment]
@chiaki-luo What do you mean by directly listening to the mel produced in the training stage and it sounds worse than inference? Maybe you can open a new issue if it is not related to this thread? What do you mean by using the output of inference as ground truth? I don't really understand your question.

**Comment by chiaki-luo** [comment]
@yl4579 I'm sorry for my poor english. What I mean is that the codes of reconstruction mel are different in train_second_TTS.py and Inference_LJSpeech.ipynb. For example, the attention of code in Inference_LJSpeech.ipynb is
`        pred_aln_trg = torch.zeros(input_lengths, int(pred_dur.sum().data))
        c_frame = 0
        for i in range(pred_aln_trg.size(0)):
            pred_aln_trg[i, c_frame:c_frame + int(pred_dur[i].data)] = 1
            c_frame += int(pred_dur[i].data)`
which is different from s2s_attn_mono in train_second_TTS.py. And the difference makes the time length also different between the mel_rec if you feed one same text to train_second_TTS.py and Inference_LJSpeech.ipynb. And I think the quality of audio in Inference_LJSpeech.ipynb is better than mel produced in train_second_TTS.py, so I'm trying to use the code in Inference_LJSpeech.ipynb repalcing the code in training. However, the different time length is hard to make a batch tensor, so I'm trying use the code above `duration = model.predictor.duration_proj(x) / speed` to control the time length to make a batch.

**Comment by bobo-paopao** [comment]
> Yes, just scale the duration predicted. Simply change the line
> 
> ```python
> duration = model.predictor.duration_proj(x)
> ```
> 
> in [Inference_LJSpeech.ipynb](https://github.com/yl4579/StyleTTS/blob/main/Demo/Inference_LJSpeech.ipynb) with
> 
> ```python
> duration = model.predictor.duration_proj(x) / speed
> ```
> 
> where `speed` is your scaling factor, for example, 1.25 is 125% faster.
> 
> You can also change the speed more naturally by giving slower or faster reference audio, which is a main selling point of StyleTTS.

我发现这样用的时候并不能达到预期的控制音频速度，因为我设置了speed为3.0速度反而更慢了

**Comment by bobo-paopao** [comment]
Oh, I found that changing the code here with `duration = torch.sigmoid(duration).sum(axis=-1) / speed`  can change the speed.

**Comment by vishalsantoshi** [comment]
A question before I dive in. 

Given that `duration = torch.sigmoid(duration).sum(axis=-1)`  is available, AND if we have a target duration,  is it possible to calculate potential duration, to calculate the  `speed` factor, which would be a ratio of target duration and   potential duration ? I have been doing a double pass ( create the audio and then get the duration ) to get to the speed factor, which seems redundant.
